SAML2/XML/Signature.hs

```diff
@@ -242,56 +246,109 @@ _verifySignatureOld pks xid doc = do
-      [x] -> return x
-      _ -> throwError SignedElementNotFound
-  sx :: HXT.XmlTree
-    <- case child "Signature" x of
-      [sx] -> return sx
-      _ -> throwError SignatureNotFoundOrEmpty
-  s@Signature{ signatureSignedInfo = si } :: Signature
-    <- case docToSAML sx of
-      Left err -> throwError . SignatureParseError $ show err
-      Right v -> pure v
-  six :: BS.ByteString
-    <- failWith SignatureCanonicalizationError
-      $ (applyCanonicalization (signedInfoCanonicalizationMethod si) (Just xpath) $ DOM.mkRoot [] [x])
-  rl :: NonEmpty.NonEmpty (Either String String)
-    <- failWith (SignatureVerifyReferenceError . (show (signedInfoReference si) <>))
-      $ mapM (`verifyReference` x) (signedInfoReference si)
-
-  when (null rl) $
-    throwError . SignatureVerifyNoReferences $ show rl
-  unless (all isRight rl) $
-    throwError . SignatureVerifyBadReferences $ show (signedInfoReference si, rl)
-  unless (elem (Right xid) rl) $
-    throwError . SignatureVerifyInputNotReferenced $ show rl
+  signedSubtree :: HXT.XmlTree
+    <- do
+        mdoc' <- liftIO . try $ fixNamespaces doc
+        doc' <- case mdoc' of
+          Right x
+            -> pure x
+          Left (err :: SomeException)
+            -> throwError . SignatureParseError $ "failed to canonicalize input: " <> show err
+        case HXT.runLA (getID xid) doc' of
+          [x] -> return x
+          _ -> throwError SignedElementNotFound
+
+  signatureElem@Signature{ signatureSignedInfo = signedInfoTyped } :: Signature
+    <- do
+        sx :: HXT.XmlTree
+          <- let child n = HXT.runLA $ HXT.getChildren HXT.>>> isDSElem n HXT.>>> HXT.cleanupNamespaces HXT.collectPrefixUriPairs
+             in case child "Signature" signedSubtree of
+            [sx] -> return sx
+            _ -> throwError SignatureNotFoundOrEmpty
+
+        case docToSAML sx of
+          Left err -> throwError . SignatureParseError $ show err
+          Right v -> pure v
+
+  signedInfoElem :: BS.ByteString
+    <- let xpath = xpathbase ++ ". | " ++ xpathbase ++ "@* | " ++ xpathbase ++ "namespace::*"
+             where
+               xpathsel t = "/*[local-name()='" ++ t ++ "' and namespace-uri()='" ++ namespaceURIString ns ++ "']"
+               xpathbase = "/*" ++ xpathsel "Signature" ++ xpathsel "SignedInfo" ++ "//"
+       in failWith SignatureCanonicalizationError
+          . capture' "applyCanonicalization"
+          $ (applyCanonicalization (signedInfoCanonicalizationMethod signedInfoTyped) (Just xpath) $ DOM.mkRoot [] [signedSubtree])
+
+  referenceChecks :: NonEmpty.NonEmpty (Either String String)
+    <- failWith (SignatureVerifyReferenceError . (show (signedInfoReference signedInfoTyped) <>))
+      . capture' "verifyReference"
+      $ mapM (`verifyReference` signedSubtree) (signedInfoReference signedInfoTyped)
+
+  unless (all isRight referenceChecks) $
+    throwError . SignatureVerifyBadReferences $ show (signedInfoReference signedInfoTyped, referenceChecks)
```

Comment:
If there's a left in the referenceChecks should we possibly `sequence` it to collect the error out? Or is it more descriptive to print everything out like this?

Reply:
It's a list of `Either`s, and I want to keep track of how many errors there are, and which precisely.

I agree it looks weird, but I can't think of a better way to do it.  Wanna try something and add a commit?

Reply:
In that case; how about `show (signedInfoReference signedInfoTyped, lefts referenceChecks)` instead? `lefts` will just grab all the errors from the list of Eithers